### ISSUE TITLE
[$1000] Android- Workspaces - Comma is disabled in the number pad for language Spanish

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#29533]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
Issue found when executing PR https://github.com/Expensify/App/pull/29070

**Version Number:** 1.3.83.4

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause-Internal Team

**Slack conversation:** @

## Action Performed:
1. Launch Expensify.
2. Navigate to Settings > Preferences and change the Language to Spanish
3. Navigate to Settings > Workspaces (Espacio de trabajo) > Reimbursements (Reembolsos) > Track distance (Medir distancia) and click on the push to page item containing the current rate.
4. Enter any rate with a comma as the decimal separator.

## Expected Result:
Comma is not disabled in the number pad for language Spanish

## Actual Result:
Comma is disabled in the number pad for language Spanish

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/115492554/8f3f55d7-0c62-40d3-ad6f-91a172f47722

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0128a0d2f68a80388a</li>
        <li>Upwork Job ID: 1712737843170938880</li>
        <li>Last Price Increase: 2024-01-11</li>
<li>Automatic offers: </li>
<ul>
<li>akinwale | Contributor | 27472369</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @johncschuster</details>

**Comment by s-alves10** [comment]
I'm not able to reproduce this

**Comment by kbecciv** [comment]
@pradeepmdk Issue is reproducible on Samsung devices ( tested on SG21/13) Thanks

**Comment by DaniloVlad** [comment]
Unable to recreate the issue locally.
Tried with both Android 13 and 12. Not sure if my environment is set up 100% as I can't access the SO link in the top-level readme which suggests important set up steps for android.

However, here is a video of Android 12.0 working:


https://github.com/Expensify/App/assets/15314885/9c1e2084-ba87-4da6-8c2c-d51b930a60ad

@johncschuster @kbecciv

**Comment by johncschuster** [comment]
Thank you for trying that, @DaniloVlad. 

@kbecciv, can you clarify what steps we need to follow to reproduce this? Is this behavior exclusive to Samsung devices?

**Comment by johncschuster** [comment]
@kbecciv bump! Can you please clarify the reproduction steps? Is this specific to Samsung devices?

**Comment by kbecciv** [comment]
Yes, this is issue reproducible on Sumsung devices, confirmed with 2 testers.

**Comment by burczu** [comment]
@johncschuster Unfortunately I have no access to any Samsung device, so I've asked on Slack if anyone is interested in taking this over and @akinwale is eager to take it, so please assign them instead of me. Thank you!

**Comment by burczu** [comment]
@johncschuster bump https://github.com/Expensify/App/issues/29533#issuecomment-1782374805

**Comment by johncschuster** [comment]
Thanks for the bump! I missed that.

**Comment by johncschuster** [comment]
Welp, Melvin isn't right here. @akinwale is the C+ Reviewer (taking over for @burczu). I'm not sure if that really matters. Asking in [Slack](https://expensify.slack.com/archives/C01SKUP7QR0/p1698849390496519).

**Comment by akinwale** [comment]
Advertised in Slack for more 👀 

https://expensify.slack.com/archives/C01GTK53T8Q/p1699365355921399

**Comment by akinwale** [comment]
> Can we consider changing "." to "," automatically when language is spanish and changing "." to "," when language is English to be new expected behaviour instead of applying a workaround to a mobile model specific issue @akinwale wdyt?

That's a workaround which isn't very elegant. Users would expect whatever characters they input to be shown on the screen. I think we can explore the option of using a virtual keyboard instead. The BigNumberPad component can be adapted to display `,` or `.` based on the current language.

**Comment by akinwale** [comment]
> This may require design team review as there is another input in for unit selection.

Yes, I noticed. However, the soft keyboard is not displayed for unit selection since it's a dropdown.

**Comment by johncschuster** [comment]
Bumped in Slack

**Comment by akinwale** [comment]
@johncschuster Any thoughts on getting opinions from the design team on using the BigNumberPad for this issue?

Also, could we consider increasing the bounty on this to get more interest?

**Comment by johncschuster** [comment]
@akinwale sorry I missed your comment above! Yes, I'm totally okay with getting the @Expensify/design team's feedback.

Let's get their feedback first, and then raise the bounty so we can get more eyes on this one.

**Comment by shawnborton** [comment]
Hmm @JmillsExpensify do you have strong thoughts here? 

I can see an argument to do:

> The BigNumberPad component can be adapted to display , or . based on the current language.

**Comment by trjExpensify** [comment]
I don't have a strong opinion on it, but I do agree if you press `.` we should show a `.`, same goes for `,`. So if we do let you enter either of those when your language is set to Spanish -- I think that's okay.

**Comment by johncschuster** [comment]
Thanks for your thoughts, @trjExpensify! That makes sense to me too!

**Comment by johncschuster** [comment]
I'm going to lean on @shawnborton and @trjExpensify for this one.

**Comment by shawnborton** [comment]
Hmm I would think that the right solution would be to switch the `.` for a `,` on the big number pad if the language is set to Spanish

**Comment by akinwale** [comment]
> Hmm I would think that the right solution would be to switch the `.` for a `,` on the big number pad if the language is set to Spanish

Just to summarise the expected behaviour here:
1. Instead of displaying the default system soft keyboard on the Track distance page, we can disable (meaning prevent from being displayed) the soft keyboard and display a BigNumberPad instead when the input field is in focus.
2. When the language is set to Spanish, change the `.` to the `,` only for the Track Distance screen.

@johncschuster @trjExpensify Anything else you think we should add here?

**Comment by trjExpensify** [comment]
>Apply the same workaround we have for ios, for spanish language there is no error if user adds "." instead of "," we simply accepts it as it is not possible to show "," programatically IOS keyboard

Interesting that the iOS keyboard doesn't support `,` instead of `.` for Spanish. 

@shawnborton if that's the case and we're switching the implementation here for this numerical only field to our BNP to get around it, I sorta' think we need to be consistent and look at that update everywhere in /app?

**Comment by shawnborton** [comment]
> I sorta' think we need to be consistent and look at that update everywhere in /app?

Yeah that was my thinking. Anywhere the app shows the BNP while in Spanish would get the switch to having a comma.

**Comment by trjExpensify** [comment]
Right, but to confirm you agree it's kinda' two-fold yeah?

- Anywhere we show a numerical keyboard in app today, we should update it to the BNP component
- Update the BNP to show a `,` instead of `.` when language is set to Spanish.

**Comment by shawnborton** [comment]
Ah interesting, I didn't realize the standard platform numerical keyboard was an issue as well. I wasn't thinking we'd need to update those views to the BNP. 

In which views where we have the default platform numerical keyboard is the comma thing an issue?

**Comment by trjExpensify** [comment]
Ohhh.. I did not glean that from above! Haha, so to summarise: 

1. On Android & iOS, the numerical keyboard shows `,` or `.` depending on your device's language settings.
2. NewDot validation for accepting `,` or `.` as a valid input is based on our app's language settings. i.e if it's set to Spanish, we expect `,`. If it's set to English, we expect `.`
3. That presents a failure case whereby somebody's device language and app language has a mismatch. For example, the device is set to `Spanish` (therefore only showing a `,` on the keyboard), and the app is set to English (therefore expecting a `.`). As a result of our validation logic being based on our app's language settings, we error out expecting a `.` when the user has inputted a `,`. 
4. This issue has come up before somewhere with iOS and we fixed it by accepting both `,` and `.` as valid inputs for the decimal place.
5. The question now is; should we do the same for Android or should we change to the BNP component to show `,` || `.` based on our app's language settings no the device language settings? 

Is that right?

**Comment by akinwale** [comment]
To clarify a couple of points:

> 1. On Android & iOS, the numerical keyboard shows `,` or `.` depending on your device's language settings.

The actual issue here is on Samsung Android devices specifically, where the soft keyboard is the default Samsung keyboard, the `,` button is greyed out on the keyboard, so the user can only input `.`, even when the device language is set to Spanish. This does not happen on other soft keyboards such as Gboard (both `,` and `.` can be input).

> 2. NewDot validation for accepting `,` or `.` as a valid input is based on our app's language settings. i.e if it's set to Spanish, we expect `,`. If it's set to English, we expect `.`

For validation on this field in particular, ND accepts either `,` or `.` without errors (there is logic to automatically convert the separator to `.` before validating using `parseFloat`).

https://github.com/Expensify/App/blob/af91f95e123e86f58e198a7b518377e2508cf12f/src/libs/NumberUtils.ts#L80-L82

> 3. That presents a failure case whereby somebody's device language and app language has a mismatch. For example, the device is set to `Spanish` (therefore only showing a `,` on the keyboard), and the app is set to English (therefore expecting a `.`). As a result of our validation logic being based on our app's language settings, we error out expecting a . when the user has inputted a `,`.

Therefore, the failure case here is that the user cannot enter `,` on a Samsung Android device using the default Samsung soft keyboard. The reason why we're suggesting using the BNP is so that we can allow the user to input their preferred decimal separator regardless.

Hope this makes sense.

**Comment by akinwale** [comment]
> > Samsung keyboard, the , button is greyed out on the keyboard, so the user can only input ., even when the device language is set to Spanish.
> 
> I am saying when the the samsung phone is booted first time it asks for region or automatically detects i am not 100% sure and based on that the "," is greyed out or not, there is no other reason samsung would grey it out
>

You can also try to change the language and region in the device settings. When I did this, the comma was still greyed out.

**Comment by trjExpensify** [comment]
>The actual issue here is on Samsung Android devices specifically, where the soft keyboard is the default Samsung keyboard, the , button is greyed out on the keyboard, so the user can only input ., even when the device language is set to Spanish.

Why?

**Comment by trjExpensify** [comment]
>This bug is from Samsung upstream

Do you have a reference or something to corroborate that I can look at? We're basically saying the native Samsung numerical keyboard features a dead/unresponsive `,` button for all apps using their native numerical keyboard input. That seems pretty wild.

**Comment by trjExpensify** [comment]
What about this? https://stackoverflow.com/a/42613077/10409556

**Comment by johncschuster** [comment]
We're still discussing this one, Melv

**Comment by johncschuster** [comment]
@abekkala I'm OOO starting Monday, December 18 (today) through Tuesday, January 2.

**Current status:**
We're discussing the best path forward

If this issue is open when I'm back from OOO, I'll take it back over. Thank you!

**Comment by abekkala** [comment]
I was assigned as the second assignee while John is ooo - but I am also ooo. 😬 
Johns current status is above. https://github.com/Expensify/App/issues/29533#issuecomment-1860851799

**Comment by johncschuster** [comment]
Thanks for taking care of this while I was OOO @abekkala and @trjExpensify. I'll take it back over.

**Comment by johncschuster** [comment]
Oh whoops! @trjExpensify is part of the discussion. Adding him back! 😂

**Comment by akinwale** [comment]
@johncschuster Any thoughts on increasing the bounty on this one for more interest?

**Comment by johncschuster** [comment]
@akinwale I'm totally open to increasing the bounty to get more eyes on this.

**Comment by shubham1206agra** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue

Android- Workspaces - Comma is disabled in the number pad for language Spanish

### What is the root cause of that problem?

From the discussion happened above, this looks like problem with platform (Samsung) instead of Expensify App code.

### What changes do you think we should make in order to solve the problem?

To not impact the Spanish user base, we can convert both input as a MenuItem. 

The first MenuItem where we input rate, we can use MoneyRequestAmountForm (not exactly, will require some modification to fit the purpose). The reason is simple: it gives a dedicated screen with BNP component, and we will not worry about such problems, and also about other input.

The second MenuItem can be simple value picker on a dedicated screen, very similar to Language Picker or Theme Picker in Preference page.

About the consistency: We are doing a similar design pattern in VBBA refactor, where we split the input across different screens, and using MenuItem to access (edit) a certain input value.

### What alternative solutions did you explore? (Optional)

**Comment by akinwale** [comment]
The suggested approach using the push-to-page paradigm in @shubham1206agra's [proposal](https://github.com/Expensify/App/issues/29533#issuecomment-1887506825) seems to be reasonable enough, as it would further make the user experience consistent across the app. This means the flow for editing the distance rate would be similar to editing the amount for a money / IOU request.

🎀👀🎀 C+ reviewed.

cc @trjExpensify @shawnborton if you have any further thoughts on this.

**Comment by tgolen** [comment]
I went back and read through all the comments, and my thoughts are:

- This issue in it's current form is very low value
- Let's minimize the amount of work we need to do to fix this and move onto other things
- I am more in favor of the same workaround we are doing for iOS (allowing them to enter a `.`)
- I want to have an app that is known for being high quality (which discounts my previous points)
- I think the most ideal outcome is to "standardise on the BNP for all numerical fields everywhere so they are mostly consistent in the UI"

In light of those thoughts, I do not want to accept the proposal from @shubham1206agra and I would like to see a proposal that implements the BNP here, then I think we should open up another GH to find all the rest of the places and implement the BNP everywhere else.

**Comment by shubham1206agra** [comment]
> In light of those thoughts, I do not want to accept the proposal from @shubham1206agra and I would like to see a proposal that implements the BNP here, then I think we should open up another GH to find all the rest of the places and implement the BNP everywhere else.

@tgolen Didn't I propose the use of BNP here? I am confused whats the problem here. Is it separate screen or layout of the page?

**Comment by shubham1206agra** [comment]
Yes. Because without new page it will look really weird, and will not be able to generalize any other places

**Comment by tgolen** [comment]
OK, you're right. I went and looked at the other places we use the BNP and I think it would take more work than I had imagined to get it to work in a non-dedicated-page scenario. In light of that, I accept your proposal.

**Comment by shubham1206agra** [comment]
> I am still wondering if the bug itself justifies the change of numerical input keyboard across the app, The flow itself is rarely used by user, most probably on initial on setup of workspace and affects a very small set of users with specific android device and its a known issue for device manufacturer.

Not all the numerical keyboard. The keyboard where amount / decimal entering is an option.

> 
> > I am more in favor of the same workaround we are doing for iOS (allowing them to enter a .)
> 
> @tgolen thats exactly what i am proposing [here](https://github.com/Expensify/App/issues/29533#issuecomment-1798804260).

**Comment by shubham1206agra** [comment]
I will try to finish this by tomorrow

**Comment by johncschuster** [comment]
Thanks @shubham1206agra! Keep us posted!

**Comment by tgolen** [comment]
Update: PR is still in the works. It looks like it's still in draft. @shubham1206agra you said you were going to try to finish it last week. Can you please post an update with an adjusted ETA?

**Comment by shubham1206agra** [comment]
@tgolen Actually I will be OOO from 24th to 29th Jan. I will try to finish this tomorrow most probably.

**Comment by tgolen** [comment]
OK, thank you for the update. I'm going to drop this down to weekly so I don't have to keep posting updates while you're OOO.

**Comment by johncschuster** [comment]
Hi @shubham1206agra, how's the PR been going? Can we get an update here when you get a moment?

**Comment by shubham1206agra** [comment]
Since Ideal Nav merged just now. I need some time to fix conflicts that will arise from Ideal Nav.

**Comment by shubham1206agra** [comment]
Update: I need https://github.com/Expensify/App/pull/34849#issuecomment-1925327605 to be addressed, and PR will be good to go for review.

**Comment by tgolen** [comment]
# Weekly Update
- [The PR](https://github.com/Expensify/App/pull/34849) is still in active development and final changes are being made. It looks like it should be ready for review soon (maybe today or Monday).

**Comment by akinwale** [comment]
Looks like automation missed this. The PR was deployed to production on 2022-02-15, so the payment due date is 2022-02-22.

cc @johncschuster

**Comment by johncschuster** [comment]
Sorry for missing this! I'll get this taken care of now.

**Comment by johncschuster** [comment]
I've extended invites to [this job](https://www.upwork.com/ab/applicants/1765831476036349952/job-details) to both @akinwale and @shubham1206agra, can you please accept that so I can issue payment? Thank you!

**Comment by johncschuster** [comment]
Thanks for everyone's patience! Coming from the [Contributing Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#regressions), payment will be subject to a regression penalty:

> If a PR causes a regression at any point within the regression period (starting when the code is merged and ending 168 hours (that's 7 days) after being deployed to production):
> 
> payments will be issued 7 days after all regressions are fixed (ie: deployed to production)
> a 50% penalty will be applied to the Contributor and [Contributor+](https://github.com/Expensify/App/blob/main/contributingGuides/HOW_TO_BECOME_A_CONTRIBUTOR_PLUS.md) for each regression on an issue

@akinwale, I'm issuing payment to you now.
@shubham1206agra, can you please ping me when you've gotten your Upwork account issue taken care of? I'll take care of payment after that.

**Comment by johncschuster** [comment]
@shubham1206agra I don't see anything in our contributing docs that makes an exception for the regression penalty based on the type of regression. Do you see anything that makes an exception? I'm happy to discuss it with the team, but without a documented exception case, I'm inclined to follow the process and apply the penalty.

**Comment by joekaufmanexpensify** [comment]
Got assigned [this one](https://github.com/Expensify/App/issues/39641). Think it may be same root cause. Curious if others agree?

**Comment by johncschuster** [comment]
# Payment Summary
@shubham1206agra - $500 to be paid via Manual Request

**Comment by johncschuster** [comment]
Hi @shubham1206agra, I'm sorry for the delay here. I was under the impression you would be paid via Manual request. Is that not correct?

**Comment by johncschuster** [comment]
Thanks for posting your Upwork profile, @shubham1206agra! I've extended the offer, [here](https://www.upwork.com/nx/wm/offer/102308832). The comment above has already been discussed, and we're going to stick with our decision.

**Comment by johncschuster** [comment]
Payment has been issued! 🎉
